rust/worker/src/execution/operators/hnsw_knn.rs

```diff
@@ -109,30 +109,57 @@ impl Operator<HnswKnnOperatorInput, HnswKnnOperatorOutput> for HnswKnnOperator {
         &self,
         input: &HnswKnnOperatorInput,
     ) -> Result<HnswKnnOperatorOutput, Self::Error> {
-        // If a filter list is supplied but it does not have anything for the segment, as it implies the data is all in the log
-        // then return an empty response.
-        if !input.allowed_ids.is_empty() && input.allowed_ids_hnsw.is_empty() {
-            return Ok(HnswKnnOperatorOutput {
-                offset_ids: vec![],
-                distances: vec![],
-            });
-        }
         let record_segment_reader = match RecordSegmentReader::from_segment(
             &input.record_segment,
             &input.blockfile_provider,
         )
         .await
         {
             Ok(reader) => reader,
+            Err(e) => match *e {
+                RecordSegmentReaderCreationError::UninitializedSegment => {
+                    tracing::error!(
+                        "[HnswKnnOperation]: Error creating record segment reader {:?}",
+                        *e
+                    );
+                    return Ok(HnswKnnOperatorOutput {
+                        offset_ids: vec![],
+                        distances: vec![],
+                    });
+                }
+                _ => {
+                    tracing::error!("[HnswKnnOperation]: Error creating record segment {:?}", e);
+                    return Err(Box::new(HnswKnnOperatorError::RecordSegmentError));
+                }
+            },
+        };
+        let log_materializer = LogMaterializer::new(
+            Some(record_segment_reader.clone()),
+            input.logs.clone(),
+            None,
+        );
+        let logs = match log_materializer.materialize().await {
+            Ok(logs) => logs,
             Err(e) => {
-                // This should not happen even for uninitialized segment because
-                // we return early in case when record segment is uninitialized.
-                tracing::error!("[HnswKnnOperation]: Error creating record segment {:?}", e);
-                return Err(Box::new(HnswKnnOperatorError::RecordSegmentError));
+                tracing::error!("[HnswKnnOperation]: Error materializing logs {:?}", e);
+                return Err(Box::new(HnswKnnOperatorError::LogMaterializationError(e)));
             }
         };
+        let mut remaining_allowed_ids: HashSet<&String> =
```

Comment:
note to self: revisit to understand

Reply:
I think we can push Adds onto the Disallowed ids() - then you may not need this allowed id filtering

Reply:
Discussed that it is not trivial